akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala

```diff
@@ -0,0 +1,170 @@
+import scala.util.{ Failure, Success, Try }
+
+@InternalApi private[akka] final class FutureFlow[In, Out, M](futureFlow: Future[Flow[In, Out, M]])
+    extends GraphStageWithMaterializedValue[FlowShape[In, Out], Future[M]] {
+  val in = Inlet[In](s"${this}.in")
+  val out = Outlet[Out](s"${this}.out")
+  override val shape: FlowShape[In, Out] = FlowShape(in, out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[M]) = {
+    val innerMatValue = Promise[M]
+    val logic = new GraphStageLogic(shape) {
+
+      //seems like we must set handlers BEFORE preStart
+      setHandlers(in, out, Initializing)
+
+      override def preStart(): Unit = {
+        futureFlow.value match {
+          case Some(tryFlow) =>
+            Initializing.onFuture(tryFlow)
+          case None =>
+            val cb = getAsyncCallback(Initializing.onFuture)
+            futureFlow.onComplete(cb.invoke)(ExecutionContexts.parasitic)
+            //in case both ports are closed before future completion
+            setKeepGoing(true)
+        }
+      }
+
+      override def postStop(): Unit = {
+        if (!innerMatValue.isCompleted) {
+          innerMatValue.failure(new AbruptStageTerminationException(this))
+        }
+      }
+
+      object Initializing extends InHandler with OutHandler {
+        override def onPush(): Unit = {
+          throw new IllegalStateException("unexpected push during initialization")
+        }
+
+        var upstreamFailure = OptionVal.none[Throwable]
+        override def onUpstreamFailure(ex: Throwable): Unit = {
+          upstreamFailure = OptionVal.Some(ex)
+        }
+
+        override def onUpstreamFinish(): Unit = {
+          //will later be propagated to the materialized flow
+        }
+
+        var hasBeenPulled = false
+        override def onPull(): Unit = {
+          hasBeenPulled = true
+        }
+
+        var downstreamCause = OptionVal.none[Throwable]
+        override def onDownstreamFinish(cause: Throwable): Unit = {
+          downstreamCause = OptionVal.Some(cause)
+        }
+
+        def onFuture(futureRes: Try[Flow[In, Out, M]]) = futureRes match {
+          case Failure(exception) =>
+            setKeepGoing(false)
+            innerMatValue.failure(new NeverMaterializedException(exception))
+            failStage(exception)
+          case Success(flow) =>
+            //materialize flow, connect to inlets, feed with potential events and set handlers
+            connect(flow)
+            setKeepGoing(false)
+        }
+      }
+      def connect(flow: Flow[In, Out, M]): Unit = {
+        val subSource = new SubSourceOutlet[In](s"${FutureFlow.this}.subIn")
+        val subSink = new SubSinkInlet[Out](s"${FutureFlow.this}.subOut")
+
+        subSource.setHandler {
+          new OutHandler {
+            override def onPull(): Unit = if (!isClosed(in)) tryPull(in)
+
+            override def onDownstreamFinish(cause: Throwable): Unit = {
```

Comment:
Make these handler methods all one-liners without braces where possible. Also further down.

Reply:
There's a few more event handlers in `Initializers` that can be made one-liners too.